### DISCUSSION TITLE
Response Model for Exceptions

### DISCUSSION THREAD

**Discussion by troyswanson** [fastapi/fastapi#8224] [Questions] [answered]
_This is a splinter conversation from #16_

The docs for FastAPI say that you should raise an exception when an error occurs in your route logic, not return an object. Per #39, it seems like an elegant way to handle errors and I like it a lot.

In following with that pattern, I wonder if it is possible to extend the `@app.exception_handler` decorator to specify a Pydantic object that will be returned, the same way a standard `@app.get` decorator would? This might make the feature requested in #16 a little cleaner -- you could specify a list of codes that a route might throw, and the API documentation generator could figure it out from there.

Example code based on the suggestion:

```python
from fastapi import FastAPI
from pydantic import BaseModel

class Item(BaseModel):
    name: str

class GenericHTTPError(BaseModel):
    status_code: int
    detail: str

items = [Item(name="foo"), Item(name='bar')]

app = FastAPI()

@app.get('/items/{item_id}', response_model=Item, errors=[404])
async def get_item(item_id: int):
    if item_id not in items:
        raise HTTPException(status_code=404, detail="Item not found")
    return {"item": items[item_id]}

@app.exception_handler(HTTPException, response_model=GenericHTTPError)
async def generic_error(request, ex):
    return GenericHTTPError(status_code=ex.status_code, detail=ex.detail)
```

I feel like I'm missing some details, but what do you think?

**Comment by tiangolo** [comment]
Thanks for the idea!

I think there are a couple of caveats to have in mind, but I'll check it with more attention. But first, let's solve #16, as that could include a solution for this too (at least an initial one).

**Comment by tiangolo** [comment, marked as answer]
It's now possible to include `responses` in *path operations*: https://fastapi.tiangolo.com/tutorial/additional-responses/

You can also include responses that apply to several *path operations*: https://fastapi.tiangolo.com/tutorial/additional-responses/#combine-predefined-responses-and-custom-ones

And it's also supported in `APIRouter`, so you can add some responses to a whole sub-router in a single `.include_router()`: https://fastapi.tiangolo.com/tutorial/bigger-applications/#add-some-custom-tags-and-responses

Of course, this includes exceptions/error responses.

It's available in version `0.12.0`.

**Comment by tiangolo** [comment]
I guess this would be solved now by that new feature of the `responses`. So I'll close this issue now. But feel free to add more comments or create new issues.

**Comment by grbrlks** [comment]
I think this would still be a nice feature as there are some limitations in regards of the `responses` feature. 

You could also pass a custom `response_class` to specify the *media type*. With the `responses` feature I am not able to specify the *media type* of a *specific response* of the *path operation*, as far as I know. 

@tiangolo please let me know if there is a way to do this or if you consider to implement this feature, keep up the good work!

**Comment by dodumosu** [comment]
> It's now possible to include `responses` in _path operations_: https://fastapi.tiangolo.com/tutorial/additional-responses/
> 
> You can also include responses that apply to several _path operations_: https://fastapi.tiangolo.com/tutorial/additional-responses/#combine-predefined-responses-and-custom-ones
> 
> And it's also supported in `APIRouter`, so you can add some responses to a whole sub-router in a single `.include_router()`: https://fastapi.tiangolo.com/tutorial/bigger-applications/#add-some-custom-tags-and-responses
> 
> Of course, this includes exceptions/error responses.
> 
> It's available in version `0.12.0`.

@tiangolo first off, thanks for creating and maintaining FastAPI!

i got here looking for some answers, but the links you shared to the documentation are currently broken

i'm trying to use a hierarchy of dependencies to get and filter out users with a specific permission, so something like ```get_user_from_bearer_token()``` returns the user account, and ```check_for_permission()``` (a parameterised dependency) depends on ```get_user_from_bearer_token()``` to check that the returned user has the passed in permission(s). both dependencies currently raise an exception, but that shows up in the documentation as being undocumented. is there a way to add them to the generated OpenAPI spec?

thanks again!

**EDIT:** i tried sticking a Pydantic model into the ```responses``` dict (of my router) for a HTTP 403 error, but that didn't work :)

**Reply by mikeedjones** [reply]
You can [add response models in the path operation functions](https://fastapi.tiangolo.com/tutorial/response-model/) - you can also [extend or overwrite the default responses](https://fastapi.tiangolo.com/advanced/additional-responses/#additional-response-with-model). 

Does that help? I'm having trouble understanding your use case - maybe if you included a minimum example?

**Reply by dodumosu** [reply]
sure. i'll skip imports

```python
security = HTTPBearer()


def get_bearer_token(credentials: HTTPAuthorizationCredentials = Depends(security)):
    return credentials.credentials


def get_user(bearer_token: Annotated[str, Depends(get_bearer_token)]):
    # some magic here that returns a User instance
    if not user:
        raise HTTPException(status=status.HTTP_403_FORBIDDEN, detail="Could not validate token")


class RequirePermission:
    def __init__(self, permission_name: str):
        self.required_permission = permission_name
   def __call__(self, user: Annotated[User, Depends(get_user)]):
        # the User has the has_perm() method defined
        if not user.has_perm(self.required_permission):
            raise HTTPException(status=status.HTTP_404_FORBIDDEN, detail="User cannot access this resource")
```

the dependencies raise a `HTTPException`, but that shows up in the rendered docs as being undocumented:
![Screenshot 2023-07-11 at 05-57-45 OneAPI - Swagger UI](https://github.com/tiangolo/fastapi/assets/1408662/2cf7cecb-39b4-4e05-8389-af4818ed2aff)
